### PR TITLE
General multiplication in Cedar

### PR DESCRIPTION
**Pull request by cdisselkoen**
[Rendered](https://github.com/cedar-policy/rfcs/blob/general-multiplication/text/0057-general-multiplication.md)

By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.

### CODE REVIEW

**Review by aaronjeline — APPROVED** [review]
Seems reasonable to me, maybe include a bit about it being easier to overflow via multiplication compared to additon

**Review by emina — APPROVED** [review]
I'm in favor of this generalization.

File: text/0057-general-multiplication.md [hunk 1]
```diff
@@ -0,0 +1,91 @@
...
+## Motivation
+
+Today, Cedar only supports multiplication of two operands if (at least) one
+operand is a constant.
+In the more general n-ary (chained) multiplication case, Cedar requires that at
+most one operand is a non-constant, but allows the constants and non-constant to
+appear in any order.
+
+- This creates some confusion for Cedar users. No other programming language
+  that I know of has this restriction on multiplication.
```
Comment (emina) [inline text/0057-general-multiplication.md:34]: Style nit:  this is the only occurrence of first person singular in the RFC.  I'd replace it or drop the sentence.

Comment (cdisselkoen) [reply text/0057-general-multiplication.md]: Fixed

**Review by andrewmwells-amazon — APPROVED** [review]
Seems like a good idea.

File: text/0057-general-multiplication.md [hunk 2]
```diff
@@ -0,0 +1,91 @@
+# General multiplication in Cedar
+
+## Related issues and PRs
+
+- Reference Issues: related to [RFC 50]
```
Comment (mwhicks1) [inline text/0057-general-multiplication.md:5]: Is there a correctness issue that this relates to, e.g, relating to overflow and ordering issues that you hint at below? If so, reference that here.

File: text/0057-general-multiplication.md [hunk 3]
```diff
@@ -0,0 +1,91 @@
...
+more expensive than multiplication by an arbitrary constant (ignoring special
+cases such as 0 or constant powers of 2).
+Some multiplication expressions will be expensive to analyze (computationally),
+but that is already true for some other kinds of Cedar expressions that don't
+involve multiplication.
+In general, we'd prefer to use a linter to warn users about (these and other)
+expensive expressions, rather than use restrictions on the Cedar language.
+
+This RFC firmly closes the door on changing Cedar's default numeric type to
+bignum. However, that door is already basically closed.
```
Comment (mwhicks1) [inline text/0057-general-multiplication.md:75]: Might mention that bigints can always be added as an extension type later.

**Review by mwhicks1 — APPROVED** [review]
LGTM

File: text/0057-general-multiplication.md (hunk 3 above)
Comment (cdisselkoen) [reply text/0057-general-multiplication.md:75]: Added a comment.

File: text/0057-general-multiplication.md [hunk 4]
```diff
@@ -0,0 +1,92 @@
...
+## Detailed design
+
+This will actually simplify the evaluator code, validator code, and internal AST.
+Multiplication will become an ordinary `BinaryOp`, and no longer need a special
+AST node.
+The evaluator and validator can handle multiplication in the same way as
+addition and in the same code paths.
+
+This RFC does not introduce any new categories of errors, and does not result in
+changed behavior for any policies that are valid today.
```
Comment (khieta) [inline text/0057-general-multiplication.md:58]: Are there any expressions for which the previous semantics will not overflow, but the new one will? One half-baked idea based on #50: I would expect an expression like `CONST * CONST * content.value` to overflow for large `CONST` and `content.value=0` under the proposed semantics. But under the old semantics, during parsing this expression is converted to `content.value * CONST * CONST` which cannot overflow.

Should we consider this a behavior change?

**Review by khieta — APPROVED** [review]
I'm in favor of this proposal, and prefer it to #50.

File: text/0057-general-multiplication.md (hunk 2 above)
Comment (cdisselkoen) [reply text/0057-general-multiplication.md:5]: There is not currently a referenceable issue, beyond conversation on RFC 50

File: text/0057-general-multiplication.md (hunk 4 above)
Comment (cdisselkoen) [reply text/0057-general-multiplication.md:58]: Unfortunately, your example is correct.  I will update the language here.

Comment (cdisselkoen) [reply text/0057-general-multiplication.md]: Revised.

**Review by john-h-kastner-aws — APPROVED** [review]

**Comment by cdisselkoen** [comment]
Given the large number of approvals, the final comment period (FCP) for this RFC is starting now, with intent to accept the RFC.

The FCP will end 2024-03-14 at noon PT / 3pm ET / 7pm UTC.  Please add comments, and especially any objections, if you have any. For more on the RFC process, see https://github.com/cedar-policy/rfcs.

**Comment by shaobo-he-aws** [comment]
Makes sense to me.

**Review by shaobo-he-aws — APPROVED** [review]

**Pull request merged**